cdap-app-fabric/src/main/java/io/cdap/cdap/internal/credential/handler/CredentialProviderHttpHandler.java

```diff
@@ -0,0 +1,409 @@
+import io.cdap.cdap.common.BadRequestException;
+import io.cdap.cdap.common.ConflictException;
+import io.cdap.cdap.common.NamespaceNotFoundException;
+import io.cdap.cdap.common.NotFoundException;
+import io.cdap.cdap.common.conf.Constants.Gateway;
+import io.cdap.cdap.common.namespace.NamespaceQueryAdmin;
+import io.cdap.cdap.internal.credential.CredentialIdentityManager;
+import io.cdap.cdap.internal.credential.CredentialProfileManager;
+import io.cdap.cdap.proto.credential.CredentialIdentity;
+import io.cdap.cdap.proto.credential.CredentialProfile;
+import io.cdap.cdap.proto.element.EntityType;
+import io.cdap.cdap.proto.id.CredentialIdentityId;
+import io.cdap.cdap.proto.id.CredentialProfileId;
+import io.cdap.cdap.proto.id.NamespaceId;
+import io.cdap.cdap.proto.security.StandardPermission;
+import io.cdap.cdap.security.spi.authorization.ContextAccessEnforcer;
+import io.cdap.http.AbstractHttpHandler;
+import io.cdap.http.HttpHandler;
+import io.cdap.http.HttpResponder;
+import io.netty.buffer.ByteBufInputStream;
+import io.netty.handler.codec.http.FullHttpRequest;
+import io.netty.handler.codec.http.HttpRequest;
+import io.netty.handler.codec.http.HttpResponseStatus;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.io.Reader;
+import java.nio.charset.StandardCharsets;
+import java.util.Optional;
+import javax.ws.rs.DELETE;
+import javax.ws.rs.GET;
+import javax.ws.rs.POST;
+import javax.ws.rs.PUT;
+import javax.ws.rs.Path;
+import javax.ws.rs.PathParam;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * {@link HttpHandler} for credential providers.
+ */
+@Singleton
+@Path(Gateway.API_VERSION_3)
+public class CredentialProviderHttpHandler extends AbstractHttpHandler  {
+  private static final Logger LOG = LoggerFactory.getLogger(CredentialProviderHttpHandler.class);
+  private static final Gson GSON = new Gson();
+
+  private final ContextAccessEnforcer accessEnforcer;
+  private final NamespaceQueryAdmin namespaceQueryAdmin;
+  private final CredentialIdentityManager credentialIdentityManager;
+  private final CredentialProfileManager credentialProfileManager;
+
+  @Inject
+  CredentialProviderHttpHandler(ContextAccessEnforcer accessEnforcer,
+      NamespaceQueryAdmin namespaceQueryAdmin,
+      CredentialIdentityManager credentialIdentityManager,
+      CredentialProfileManager credentialProfileManager) {
+    this.accessEnforcer = accessEnforcer;
+    this.namespaceQueryAdmin = namespaceQueryAdmin;
+    this.credentialIdentityManager = credentialIdentityManager;
+    this.credentialProfileManager = credentialProfileManager;
+  }
+
+  /**
+   * Returns a list of supported credential providers.
+   *
+   * @param request The HTTP request.
+   * @param responder The HTTP responder.
+   */
+  @GET
+  @Path("/credentials/providers")
+  public void listProviders(HttpRequest request, HttpResponder responder) {
+    responder.sendString(HttpResponseStatus.METHOD_NOT_ALLOWED,
```

Comment:
do we need an access enforcer this for this API?

Reply:
I don't believe so, since it's just listing the available provider extensions once those are implemented.